tla/Traceccfraft.tla

```diff
@@ -168,6 +168,19 @@ TraceConflictAppendEntriesRequestNoConflictAppendEntriesRequest ==
                 \cdot NoConflictAppendEntriesRequest(m.dest, m.source, m)  
             /\ UNCHANGED <<candidateVars, leaderVars>>
 
+TruncateLog(i) ==
+    \* \* raft.h: recv_append_entries > become_aware_of_new_term > become_follower > rollback
+    /\ log[i] # <<>> 
+    /\ commitIndex[i] < Len(log[i])
+    /\ log' = [log EXCEPT ![i] = SubSeq(log[i], 1, commitIndex[i])]
```

Comment:
That's not quite what the implementation does, instead it's rollback(last_committable_index()) (raft.h:1821).

Rolling back to the last locally known committed break user-observable persistence: the primary will advertise entries as committed as soon as it has seen a quorum of ack, but there is a further delay until all potential candidates are aware of the commit. If there is an election in that window, the new node may have committable entries that were once advertised as committed, but which it itself doesn't know are committed.

If these entries are rolled back, by even one node, we may be destroying a quorum where there once was one, which was advertised to users.

Reply:
```suggestion
    /\ log' = [log EXCEPT ![i] = SubSeq(log[i], 1, MaxCommittableIndex(log[i])]
```

I think this should be ^

Reply:
Still investigating but `fancy_election.2` no longer validates with your proposed change (`MaxCommittableIndex(log[i])`).

Reply:
I agree with @achamayou, nodes should never rollback logs passed the last signature unless they have diverged from the current leader. This could lead to a safety violation down the line

Reply:
@lemmy it seems to me that there are two ways forward then:

1. log committable_indices, as you did in #5238 (sorry for commenting there before reading this)
2. update the spec to maintain committable indices per node, and clear it post election like the implementation does

They seem about equivalent to me (1 seems a little weaker, but not to a worrying extent), and 1. is easier/already done in #5238. @heidihoward what do you think?

Reply:
The interactions between the various indices is rather intricate, which is why my intuition suggests aligning the specification with the implementation.  In other words, I'd suggest to model `committable_indices` to verify that the spec's properties hold.

Reply:
I'm ok with `committable_indices` being added to the spec